src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -7925,43 +7961,67 @@ private BoundExpression BindIndexedPropertyAccess(SyntaxNode syntax, BoundExpres
+                            param.RefKind == RefKind.None)
                         {
                             CheckImplicitThisCopyInReadOnlyMember(receiverOpt, lengthOrCountProperty.GetMethod, diagnostics);
                             // note: implicit copy check on the indexer accessor happens in CheckPropertyValueKind
-                            patternIndexerAccess = new BoundIndexOrRangePatternIndexerAccess(
-                                syntax,
-                                receiverOpt,
-                                lengthOrCountProperty,
-                                property,
-                                BindToNaturalType(argument, diagnostics),
-                                property.Type);
+                            patternSymbol = property;
+                            returnType = property.Type;
+                            checkWellKnown(WellKnownMember.System_Index__GetOffset);
+                            success = true;
                             break;
                         }
                     }
                 }
             }
-            else if (receiverType.SpecialType == SpecialType.System_String)
+            else if (TryFindImplicitRangeIndexerPattern(syntax, receiverOpt, receiverType,
+                    out MethodSymbol? sliceMethod, diagnostics, lookupResult, ref discardedUseSiteInfo))
+            {
+                patternSymbol = sliceMethod;
+                returnType = sliceMethod.ReturnType;
+                CheckImplicitThisCopyInReadOnlyMember(receiverOpt, lengthOrCountProperty.GetMethod, diagnostics);
+                success = true;
+            }
+
+            lookupResult.Free();
+            return success;
+
+            void checkWellKnown(WellKnownMember member)
+            {
+                // Check required well-known member. They may not be needed
+                // during lowering, but it's simpler to always require them to prevent
+                // the user from getting surprising errors when optimizations fail
+                _ = GetWellKnownTypeMember(member, diagnostics, syntax: syntax);
+            }
+        }
+
+        private bool TryFindImplicitRangeIndexerPattern(
+            SyntaxNode syntax, BoundExpression? receiverOpt,
+            TypeSymbol receiverType,
+            [NotNullWhen(true)] out MethodSymbol? patternSymbol,
+            BindingDiagnosticBag diagnostics, LookupResult lookupResult,
+            ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            bool success = false;
+            patternSymbol = null;
+
+            if (receiverType.SpecialType == SpecialType.System_String)
             {
-                Debug.Assert(argIsRange);
                 // Look for Substring
                 var substring = (MethodSymbol)Compilation.GetSpecialTypeMember(SpecialMember.System_String__Substring);
                 if (substring is object)
```

Comment:
This should really be calling the Binder version of GetSpecialTypeMember, which will correctly report a missingmember error if missing.

Reply:
@alrz this isn't addressed.

Reply:
See `ListPattern_MissingMembers_Substring`. 

I think it is deliberate for indexer access, we give a bad arg error as if the method didn't exist, which it doesn't.

For list patterns we never report an exact missing member, we just report an unsupported type. This is consistent for  a missing Substring a well as verified in the test above.

Reply:
This case doesn't seem meaningfully different than the well-known checks we do right below if `substring` is present. I think we should report the missing member error.

Reply:
Since this is existing logic for ranges and that may be deliberate, I will add a prototype comment to revise later.